W-16429796 Create flex-conn-replica-caching.adoc (#698)
mulesoft/docs-gateway: gateway/1.9/modules/ROOT/pages/flex-conn-replica-caching.adoc

= Configuring Flex Gateway Caching in Connected Mode
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> Enable Flex Gateway configuration caching to have a Flex Gateway cache its gateway configuration in Redis shared storage. When configured, Flex replicas use the cached gateway configuration instead of each replica directly downloading the configuration from Anypoint Platform. Caching configuration reduces scale-up time and increases resiliency.

Reviewer: Flip first sentence?

> Enable Flex Gateway configuration caching to have a Flex Gateway cache its gateway configuration in Redis shared storage. When configured, Flex replicas use the cached gateway configuration instead of each replica directly downloading the configuration from Anypoint Platform. Caching configuration reduces scale-up time and increases resiliency.

Reviewer: ```suggestion
Enable Flex Gateway configuration caching to have a Flex Gateway cache its gateway configuration in Redis shared storage. Flex replicas then use the cached gateway configuration instead of each replica directly downloading the configuration from Anypoint Platform. Caching configuration reduces scale-up time and increases resiliency.
```

Reviewer: Don't use When for after, once, or if.

Flex Gateway configuration caching works by:

> . When the first Flex replica starts, the replica downloads the configuration from Anypoint Platform, encrypts the configuration, and stores it in Redis.

Reviewer: ```suggestion
. After the first Flex replica starts, the replica downloads the configuration from Anypoint Platform, encrypts the configuration, and stores it in Redis.
```

> . When other replicas start, they use the locally cached configuration rather than downloading from Anypoint Platform

Reviewer: ```suggestion
. After subsequent replicas start, they use the locally cached configuration rather than downloading from Anypoint Platform.
```

> . When the Flex Gateway's configuration is updated on Anypoint Platform, any one of the replicas downloads the necessary configuration from Anypoint Platform and stores the configuration in Redis for immediate use by the other replicas.

Reviewer: ```suggestion
. If the Flex Gateway configuration changes on Anypoint Platform, the replicas download the necessary configuration from Anypoint Platform and store it in Redis for immediate use by the other replicas.
```

> If Redis errors occur when caching is configured, each replica downloads configurations directly from Anypoint Platform. If Flex Gateway disconnects from Anypoint Platform, Flex gateway uses the cached configuration until the connections is restored.

Reviewer: ```suggestion
If caching is configured and Redis errors occur, each replica downloads configurations directly from Anypoint Platform. If Flex Gateway disconnects from Anypoint Platform, Flex gateway uses the cached configuration until connections are restored.
```

= Configuring Replica Configuration Caching in Connected Mode
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> Enable replica configuration caching to have a Flex Gateway replica cache its gateway configuration in Redis shared storage. When configured, new replicas use the cached gateway configuration instead of downloading the configuration from Anypoint Platform. Caching a replica's configuration reduces scale-up time and increases resiliency.

Reviewer 2: Instead of each replica downloading the configuration from Anypoint directly

Replica caching works by:

> . When the first Flex Replica is created, it sees that their is no configuration in the Redis shared storage.

Reviewer 2: it sees that **there** is  (typo)

> . The first replica creates the group ID and downloads its configuration from Anypoint Platform. The replica then stores the configuration in Redis.

Reviewer 2: I don't think the customer needs to know anything about a 'group ID'. This is an internal implementation detail.

. When a new Flex Replica is created, it uses the stored configuration.
. At their start up, all Flex Replicas connect to Anypoint Platform and send configuration update requests to the same group ID.

> . When the configuration is updated on Anypoint Platform, only one replica receives the configuration updates.

Reviewer 2: "only one replica receives the configuration updates"
This is an internal implementation detail that is not needed here

Reviewer 2: we can just say that updates are downloaded and stored in redis for all replicas to use

. The replica download the nec4essary configuration from Anypoint Platform and stores the configuration in Redis for the other replicas to use.

== Enable Replica Caching

. Ensure the Redis shared storage is configured.
+
To learn more, see xref:flex-conn-shared-storage-config.adoc[].

> . Enable the `FLEX_CONTROL_NODE_ENABLED=true` environment variable.

Reviewer 3: environment variable was changed to `FLEX_DATASOURCE_CONTROL_NODE_ENABLED`
